plutus-core/plutus-ir/src/PlutusIR/Transform/Inline/CallSiteInline.hs

```diff
@@ -60,120 +66,140 @@ becomes
-    -- inlining, and it won't make it any worse, so we could consider accepting this.
-    False
-
--- | Consider whether to inline a saturated application.
-considerInlineSat :: forall tyname name uni fun ann. InliningConstraints tyname name uni fun
-    => Term tyname name uni fun ann -- ^ The `body` of the `Let` term.
-    -> InlineM tyname name uni fun ann (Term tyname name uni fun ann)
-considerInlineSat tm = do
-    -- collect all the arguments of the term being applied to
-    case splitApplication tm of
-      -- if it is a `Var` that is being applied to, check to see if it's fully applied
-      (Var _ann name, ctx) -> do
-        maybeVarInfo <- gets (lookupVarInfo name)
-        case maybeVarInfo of
-          Just varInfo -> do
-            let body = varBody varInfo
-                defAsInlineTerm = varDef varInfo
-                inlineTermToTerm :: InlineTerm tyname name uni fun ann
-                  -> Term tyname name uni fun ann
-                inlineTermToTerm (Done (Dupable var)) = var
-                def = inlineTermToTerm defAsInlineTerm
-                fullyApplied = isFullyApplied (arity varInfo) ctx
-                -- It is the body that we will be left with in the program after we have
-                -- reduced the saturated application, so the size increase we will be left
-                -- with comes from the body, and that is what we need to check is okay
-                bodySizeOk = sizeIsAcceptable body
+-- | Fully apply the RHS of the given variable to the given arguments, and beta-reduce
+-- the application, if possible.
+fullyApplyAndBetaReduce ::
+  forall tyname name uni fun ann.
+  (InliningConstraints tyname name uni fun) =>
+  -- | The variable
+  VarInfo tyname name uni fun ann ->
+  -- | The arguments
+  AppContext tyname name uni fun ann ->
+  InlineM tyname name uni fun ann (Maybe (Term tyname name uni fun ann))
+fullyApplyAndBetaReduce info args0 = do
+  rhsBody <- liftDupable (let Done rhsBody = varRhsBody info in rhsBody)
+  let go ::
+        Term tyname name uni fun ann ->
+        Arity tyname name ->
+        AppContext tyname name uni fun ann ->
+        InlineM tyname name uni fun ann (Maybe (Term tyname name uni fun ann))
+      go acc arity args = case (arity, args) of
+        -- success
+        ([], _) -> pure . Just $ fillAppContext acc args
+        (TermParam param : arity', TermAppContext arg _ args') -> do
+          safe <- safeToBetaReduce param arg
+          if safe
+            then do
+              acc' <-
+                termSubstNamesM
+                  (\n -> if n == param then Just <$> PLC.rename arg else pure Nothing)
+                  acc
+              go acc' arity' args'
+            else pure Nothing
+        (TypeParam param : arity', TypeAppContext arg _ args') -> do
+          acc' <-
+            termSubstTyNamesM
+              (\n -> if n == param then Just <$> PLC.rename arg else pure Nothing)
+              acc
+          go
```

Comment:
Is this some fourmolu thing? Seems weird to break this since it clearly fits on one line...

Reply:
Good catch; that was me.